specification.md

```diff
@@ -2852,6 +2872,163 @@ CustomMessage {
 }
 ```
 
+### AvailableComponents Message
+
+Status: [Development]
+
+Each Agent may be composed of multiple sub-components. These components may have
+their own associated metadata. The AvailableComponents message allows the Agent
+to inform the OpAMP server about which components the Agent contains.
+
+The AvailableComponents message has the following structure:
+
+```protobuf
+message AvailableComponents {
+    map<string, ComponentDetails> components = 1;
+    bytes hash = 2;
+ }
+```
+
+#### AvailableComponents.components
+
+The components field contains a map of a unique ID
+to a [ComponentsDetails](#componentdetails-message) message.
+This field may be omitted from the message if the OpAMP server has not explicitly
+requested it by setting the ReportAvailableComponents flag in the preceding
+ServerToAgent message.
+
+##### Examples
+
+###### OpenTelemetry Collector
+
+Here is an example of how ComponentDetails could hold information regarding the included
+components for a custom build of [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector):
+
+```jsonc
+{
+  "receivers": {
+    "sub_component_map": {
+      "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver@v0.107.0": {
+        "metadata": {
+          "type": "hostmetrics",
+        }
+      }
+    }
+  },
+  "processors": {
+    "sub_component_map": {
+      "go.opentelemetry.io/collector/processor/batchprocessor@v0.107.0": {
+        "metadata": {
+          "type": "batch",
+        }
+      },
+      "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor@v0.107.0": {
+        "metadata": {
+          "type": "transform",
+        }
+      },
+    }
+  },
+  "exporters": {
+    "sub_component_map": {
+      "go.opentelemetry.io/collector/exporter/nopexporter@v0.107.0": {
+        "metadata": {
+          "type": "nop",
+        }
+      }
+    }
+  }
+  // ... Component list continues for extensions and collectors ...
+}
+```
+
+###### Fluent Bit
+
+Here's an example of how Fluent Bit could report what components it has available.
+
+```jsonc
+{
+  "input": {
+    "sub_component_map": {
+      "tail": {}
+    }
+  },
+  "parser": {
+    "sub_component_map": {
+      "json": {}
+    }
+  },
+  "filter": {
+    "sub_component_map": {
+      "lua": {},
+      "modify": {}
+    }
+  },
+  "output": {
+    "sub_component_map": {
+      "null": {},
+      "file": {},
+    }
+  }
+}
+```
+
+#### AvailableComponents.hash
+
+The agent-calculated hash of the components field. The agent MUST include this hash
+if it has the ability to report the components it has available.
+
+#### Initial Handshake
```

Comment:
From my reading of the issue, it sounds like the flow should optionally start with the agent sending the server an AvailableComponents message with just the hash to save a round-trip.

Reply:
:+1: I meant to imply that here, but I don't think I explicitly stated it. I've put a sentence here that makes it clearer.